doc/topics/configuration.rst
.. _topics-configuration:

=============================
Configuration file for Tryton
=============================

The configuration file controls some aspects of the behavior of Tryton.
The file uses a simple ini-file format. It consists of sections, led by a
``[section]`` header and followed by ``name = value`` entries:

.. highlight:: ini

::

    [database]
    uri = postgresql://user:password@localhost/
    path = /var/lib/trytond

For more information see ConfigParser_.

.. _ConfigParser: http://docs.python.org/2/library/configparser.html

The default value of any option can be changed using environment variables
with names using this syntax: ``TRYTOND_<SECTION>__<NAME>``.

Sections
========

This section describes the different main sections that may appear in a Tryton
configuration file, the purpose of each section, its possible keys, and their
possible values.
Some modules could request the usage of other sections for which the guideline
asks them to be named like their module.

web
---

Defines the behavior of the web interface.

listen
~~~~~~

Defines the couple of host (or IP address) and port number separated by a colon
to listen on.

Default ``localhost:8000``

hostname
~~~~~~~~

Defines the hostname to use when generating a URL when there is no request
context available, for example during a cron job.

root
~~~~

Defines the root path served by ``GET`` requests.

Default: Under the ``www`` directory of user's home running ``trytond``.

num_proxies
~~~~~~~~~~~

The number of proxy servers in front of ``trytond``.

Default: 0

cache_timeout
~~~~~~~~~~~~~

The cache timeout in seconds.

Default: 12h

cors
~~~~

The list (one per line) of origins allowed for `Cross-Origin Resource sharing
<https://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_.

avatar_base
~~~~~~~~~~~

The base URL without a path for avatar URL.

Default: ``''``

.. note:: It can be used to setup a CDN.


avatar_timeout
~~~~~~~~~~~~~~

The time in seconds that the avatar can be stored in cache.

Default: 7 days

database
--------

Defines how the database is managed.

uri
~~~

Contains the URI to connect to the SQL database. The URI follows the RFC-3986_.
The typical form is:

    database://username:password@host:port/?param1=value1&param2=value2

The parameters are database dependent, check the database documentation for a
list of valid parameters.

Default: The value of the environment variable ``TRYTOND_DATABASE_URI`` or
``sqlite://`` if not set.

The available databases are:

PostgreSQL
**********

``psycopg2`` supports two type of connections:

    - TCP/IP connection: ``postgresql://user:password@localhost:5432/``
    - Unix domain connection: ``postgresql://username:password@/``

Please refer to `psycopg2 for the complete specification of the URI
<https://www.psycopg.org/docs/module.html#psycopg2.connect>`_.

A list of parameters supported by PostgreSQL can be found in the
`documentation <https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-PARAMKEYWORDS>`_.

.. note::
   ``fallback_application_name`` parameter from aforementioned documentation can
   be set directly thanks to the ``TRYTOND_APPNAME`` environment variable.

SQLite
******

The URI is defined as ``sqlite://``

If the name of the database is ``:memory:``, the parameter ``mode`` will be set
to ``memory`` thus using a pure in-memory database.

The recognized query parameters can be found in SQLite's
`documentation <https://www.sqlite.org/uri.html#recognized_query_parameters>`_.

path
~~~~

The directory where Tryton stores files and so the user running ``trytond``
must have write access on this directory.

Default: The ``db`` folder under the user home directory running ``trytond``.

list
~~~~

A boolean value to list available databases.

Default: ``True``

retry
~~~~~

The number of retries when a database operational error occurs during a request.

Default: ``5``

language
~~~~~~~~

The main language of the database that will be used for storage in the main
table for translations.

Default: ``en``

avatar_filestore
~~~~~~~~~~~~~~~~

This configuration value indicates whether the avatars should be stored in the
:py:mod:`trytond.filestore` (``True``) or the database (``False``).

Default: ``False``

avatar_prefix
~~~~~~~~~~~~~

The prefix to use with the :ref:`FileStore <ref-filestore>` to store avatars.

Default: ``None``

default_name
~~~~~~~~~~~~

The name of the database to use for operations without a database name.
Default: ``template1`` for PostgreSQL, ``:memory:`` for SQLite.

unaccent_function
~~~~~~~~~~~~~~~~~

The name of the unaccent function.

Default: ``unaccent``

similarity_function
~~~~~~~~~~~~~~~~~~~

The name of the similarity function.

Default: ``similarity``

request
-------

max_size
~~~~~~~~

The maximum size in bytes of unauthenticated request (zero means no limit).

Default: 2MB

max_size_authenticated
~~~~~~~~~~~~~~~~~~~~~~

The maximum size in bytes of an authenticated request (zero means no limit).

Default: 2GB


cache
-----

Defines size of various cache.

model
~~~~~

The number of different model kept in the cache per transaction.

Default: ``200``

record
~~~~~~

The number of record loaded kept in the cache of the list.
It can be changed locally using the ``_record_cache_size`` key in
:attr:`Transaction.context`.

Default: ``2000``

field
~~~~~

The number of field to load with an ``eager`` :attr:`Field.loading`.

Default: ``100``

clean_timeout
~~~~~~~~~~~~~

The minimum number of seconds between two cleanings of the cache.
If the value is 0, the notification between processes will be done using
channels if the back-end supports them.

Default: ``300``

rpc.<model>.<method>
~~~~~~~~~~~~~~~~~~~~

The duration in seconds of the cache for the RPC answer to `method` of the
model named `model`. It will override the value defined in the model.

count_timeout
~~~~~~~~~~~~~

The cache timeout duration in seconds of the estimation of records.

Default: ``86400`` (1 day)

count_clear
~~~~~~~~~~~

The number of operations after which the counting estimation of records is
cleared.

Default: ``1000``

queue
-----

worker
~~~~~~

Activate asynchronous processing of the tasks. Otherwise they are performed at
the end of the requests.

Default: ``False``

clean_days
~~~~~~~~~~

The number of days after which processed tasks are removed.

Default: ``30``

table
-----

This section allows to override the default generated table name for a
:class:`ModelSQL`. The main goal is to bypass limitation on the name length of
the database backend.
For example::

    [table]
    account.invoice.line = acc_inv_line
    account.invoice.tax = acc_inv_tax

ssl
---

Activates SSL_ on the web interface.

.. note:: It is recommended to delegate the SSL support to a proxy.

privatekey
~~~~~~~~~~

The path to the private key.

certificate
~~~~~~~~~~~

The path to the certificate.

.. tip::
   Set only one of ``privatekey`` or ``certificate`` to ``true`` if the SSL is
   delegated.

email
-----

.. note:: Email settings can be tested with the ``trytond-admin`` command

uri
~~~

The SMTP-URL_ to connect to the SMTP server which is extended to support SSL_
and STARTTLS_.
The available protocols are:

    - ``smtp``: simple SMTP
    - ``smtp+tls``: SMTP with STARTTLS
    - ``smtps``: SMTP with SSL

The uri accepts the following additional parameters:

* ``local_hostname``: used as FQDN of the local host in the HELO/EHLO commands,
  if omited it will use the value of ``socket.getfqdn()``.
* ``timeout``: A number of seconds used as timeout for blocking operations. A
  ``socket.timeout`` will be raised when exceeded. If omited the default timeout
  will be used.


Default: ``smtp://localhost:25``

from
~~~~

Defines the default ``From`` address (using RFC-822_) for emails sent by Tryton.

For example::

    from: "Company Inc" <info@example.com>

session
-------

authentications
~~~~~~~~~~~~~~~

A comma separated list of the authentication methods to try when attempting to
verify a user's identity. Each method is tried in turn, following the order of
the list, until one succeeds. In order to allow `multi-factor authentication`_,
individual methods can be combined together using a plus (``+``) symbol.

Example::

    authentications = password+sms,ldap

By default, Tryton only supports the ``password`` method.  This method compares
the password entered by the user against a stored hash of the user's password.
Other modules can define additional authentication methods, please refer to
their documentation for more information.

Default: ``password``

max_age
~~~~~~~

The time in seconds that a session stay valid.

Default: ``2592000`` (30 days)

timeout
~~~~~~~

The time in seconds without activity before the session is no more fresh.

Default: ``300`` (5 minutes)

max_attempt
~~~~~~~~~~~

The maximum authentication attempt before the server answers unconditionally
``Too Many Requests`` for any other attempts. The counting is done on all
attempts over a period of ``timeout``.

Default: ``5``

max_attempt_ip_network
~~~~~~~~~~~~~~~~~~~~~~

The maximum authentication attempt from the same network before the server
answers unconditionally ``Too Many Requests`` for any other attempts. The
counting is done on all attempts over a period of ``timeout``.

Default: ``300``

ip_network_4
~~~~~~~~~~~~

The network prefix to apply on IPv4 address for counting the authentication
attempts.

Default: ``32``

ip_network_6
~~~~~~~~~~~~

The network prefix to apply on IPv6 address for counting the authentication
attempts.

Default: ``56``

password
--------

length
~~~~~~

The minimal length required for the user password.

Default: ``8``

forbidden
~~~~~~~~~

The path to a file containing one forbidden password per line.

entropy
~~~~~~~

The ratio of non repeated characters for the user password.

Default: ``0.75``

reset_timeout
~~~~~~~~~~~~~

The time in seconds until the reset password expires.

Default: ``86400`` (24h)

passlib
~~~~~~~

The path to the `INI file to load as CryptContext
<https://passlib.readthedocs.io/en/stable/narr/context-tutorial.html#loading-saving-a-cryptcontext>`_.
If not path is set, Tryton will use the schemes ``bcrypt`` or ``pbkdf2_sha512``.

Default: ``None``

attachment
----------

Defines how to store the attachments

filestore
~~~~~~~~~

A boolean value to store attachment in the :ref:`FileStore <ref-filestore>`.

Default: ``True``

store_prefix
~~~~~~~~~~~~

The prefix to use with the ``FileStore``.

Default: ``None``

bus
---

allow_subscribe
~~~~~~~~~~~~~~~

A boolean value to allow clients to subscribe to bus channels.

Default: ``False``

url_host
~~~~~~~~

If set redirects bus requests to the host URL.

long_polling_timeout
~~~~~~~~~~~~~~~~~~~~

The time in seconds to keep the connection to the client opened when using long
polling for bus messages

Default: ``300``

cache_timeout
~~~~~~~~~~~~~

The number of seconds a message should be kept by the queue before being
discarded.

Default: ``300``

select_timeout
~~~~~~~~~~~~~~

The timeout duration of the select call when listening on a channel.

Default: ``5``

html
----

src
~~~

The URL pointing to `TinyMCE <https://www.tiny.cloud/>`_ editor.

Default: ``https://cloud.tinymce.com/stable/tinymce.min.js``

plugins
~~~~~~~

The space separated list of TinyMCE plugins to load.
It can be overridden for specific models and fields using the names:
``plugins-<model>-<field>`` or ``plugins-<model>``.

Default: ``

css
~~~

The JSON list of CSS files to load.
It can be overridden for specific models and fields using the names:
``css-<model>-<field>`` or ``css-<model>``.

Default: ``[]``

class
~~~~~

The class to add on the body.
It can be overridden for specific models and fields using the names:
``class-<model>-<field>`` or ``class-<model>``.

Default: ``''``

wsgi middleware
---------------

The section lists the `WSGI middleware`_ class to load.
Each middleware can be configured with a section named ``wsgi <middleware>``
containing ``args`` and ``kwargs`` options.

Example::

    [wsgi middleware]
    ie = werkzeug.contrib.fixers.InternetExplorerFix

    [wsgi ie]
    kwargs={'fix_attach': False}


.. _JSON-RPC: http://en.wikipedia.org/wiki/JSON-RPC
.. _XML-RPC: http://en.wikipedia.org/wiki/XML-RPC
.. _RFC-3986: http://tools.ietf.org/html/rfc3986
.. _SMTP-URL: http://tools.ietf.org/html/draft-earhart-url-smtp-00
.. _RFC-822: https://tools.ietf.org/html/rfc822
.. _SSL: http://en.wikipedia.org/wiki/Secure_Sockets_Layer
.. _STARTTLS: http://en.wikipedia.org/wiki/STARTTLS
.. _WSGI middleware: https://en.wikipedia.org/wiki/Web_Server_Gateway_Interface#Specification_overview
.. _`multi-factor authentication`: https://en.wikipedia.org/wiki/Multi-factor_authentication


CHANGELOG
* Use singleton for TableHandler
* Add configuration entries for unaccent and similarity functions
* Add more context key to clean when adding to cache
* Add a way to register other objects in the pool
* Catch get_connection errors in Cache listener
* Compute cron job scheduling using server timezone
* Reset less often the user sessions
* Add option to strip leading and trailing whitespace from fields.Char
* Add database timeout parameter on RPC calls
* Combine search and get_keys in DictSchemaMixin

Version 6.4.3 - 2022-07-01
* Bug fixes (see mercurial logs for details)
* Enforce certificate validation for SMTP connection (issue11564)

Version 6.4.2 - 2022-06-15
* Bug fixes (see mercurial logs for details)

Version 6.4.1 - 2022-06-03
* Bug fixes (see mercurial logs for details)

Version 6.4.0 - 2022-05-02
* Bug fixes (see mercurial logs for details)
* Use unittest discover
* Add module tutorial
* Test all XML view and SVG icon files are used
* Add notification message
* Add validate_fields to ModelStorage
* Drop support for PyPy and psycopg2cffi
* Use a set for field's depends
* Relax the constraint on a field's depends
* Include only needed fields when fetching a view definition
* Make warning check re-entrant
* Manage DatabaseDataError in ModelSQL
* Remove support for pysqlite2
* Do not import empty One2Many data
* Deduplicate entries in domain inversion's simplify
* Prevent to create and delete singleton
* Allow CORS on root path
* Allow button access to be deactivated
* Always return tuple for MultiSelection
* Use default selectors instead of select
* Add batch option to push queue tasks
* Support ref value for reference field
* Enforce ref model in XML data
* Do not resolve entities by default with lxml (issue11219)
* Use defusedxml to parse XML (issue11244)
* Use dictionary as search_order on Reference field
* Retry sending email on temporary failure
* Order not sorted Selection by index definition
* Add optional column on tree view
* Use dictionary as domain on Reference field
* Include model, field and column in import data error message
* Support limit and offset to ModelSQL count search and search_count
* Apply view inheritance to board
* Add RPC view_get method to View
* Apply view inheritance to all matching elements
* Limit board action domain to active id and ids
* Remove entropy check on password
* Set view_id in the context when calling view_attributes
* Add ip_address and device_cookie login method options
* Add support for Python 3.10
* Remove support for Python 3.6
* Add creatable attribute on tree and form views
* Fill local cache of Function field instances with values
* Cache Function field getter without context in readonly transactions
* Use expire delay instead of UTC datetime in reset password email
* Use RENAME and DROP COLUMN with SQLite backend
* Support window functions with SQLite backend
* Use JSONB to store Dict field on PostgreSQL backend

Version 6.2.0 - 2021-11-01
* Bug fixes (see mercurial logs for details)
* Allow to cast to date with timezone
* Support create/delete attribute in view
* Skip default values when copying records
* Use local cache for relation fields in readonly transactions
* Add format method to DictSchemaMixin
* Allow modify record name on the reports
* Add methods to format number and symbol on Lang
* Add format number with symbol on Report
* Allow locking tables by calling ModelSQL.lock without records
* Add test on domain of relation fields
* Add stored path as alternative to MPTT
* Activate and upgrade modules from configuration wizard
* Add set_from_header in email tools
* Add grouping attribute to numeric widgets
* Ensure with a test that objects in the Pool have __slots__
* Use bigdecimal tag for XML-RPC
* Use tuple in Dict value instead of list
* Do not set record name on title of report get_email
* Add node to id of status
* Use ImmutableDict for Transaction context
* Use UNION for 'Or'-ed domain with subqueries
* Add remove_forbidden_chars in tools
* Manage errors during non-interactive operations
* Add estimation count to ModelStorage
* Support empty digits in format_number of Report
* Add digits mixin
* Use argon2 or scrypt to hash password by default
* Allow PYSON Expression as key for PYSON In with dict object
* Add method to format warning name with records
* Add option to trytond-console to start a readonly transaction
* Add support for werkzeug 2.0
* Combine search and get_keys in DictSchemaMixin
* Make language code unique
* Support base64 encoded data in ModelStorage.import_data
* Add BOOL_AND and BOOL_OR to SQLite backend

Version 6.0.0 - 2021-05-03
* Bug fixes (see mercurial logs for details)
* Allow column sql types to be tested from the table handler
* Add support for database connection parameters to configuration URI
* Use immutable datastructures for Dict and MultiSelection fields
* Skip warnings for non-interactive operations
* Check rule only if _check_access is set
* Add statistics to Cache
* Add support for avatars
* Add status command
* Add document widget on form view
* Add full text and similarity search and FullText field
* Add URL type to icon
* Use instance method to get next sequence value
* Store the sequence type directly on sequence
* Allow sending email with record's attachments
* Add which records to use for actions
* Add parent to group to inherit accesses
* Add __access__ to Model
* Add route wrapper to allow null origin
* Add header grouping on report
* Protect trusted devices against brute force attack
* Make ModelView.parse_view public
* Add help for each value to selection and multiselection
* Use safe_join in SharedDataMiddlewareIndex (issue10068)
* Use menu name for window opened from menu
* Add trigonometric functions to sqlite backend
* Allow skipping user warnings globally
* Add validate option to trytond-admin
* Refresh pool of other processes
* Add clear_all method to Cache
* Support Genshi element directives in HTML editor
* Add firstline tool
* Add support for Python 3.9

trytond/config.py
# This file is part of Tryton.  The COPYRIGHT file at the top level of
# this repository contains the full copyright notices and license terms.
import configparser
import logging
import os
import urllib.parse

import __main__ as main

from . import status

__all__ = ['config', 'get_hostname', 'get_port', 'split_netloc',
    'parse_listen', 'parse_uri']
logger = logging.getLogger(__name__)

# Needed so urlunsplit to always set netloc
for backend_name in ['postgresql', 'sqlite']:
    if backend_name not in urllib.parse.uses_netloc:
        urllib.parse.uses_netloc.append(backend_name)


def get_hostname(netloc):
    if '[' in netloc and ']' in netloc:
        return netloc.split(']')[0][1:]
    elif ':' in netloc:
        return netloc.split(':')[0]
    else:
        return netloc


def get_port(netloc):
    netloc = netloc.split(']')[-1]
    return int(netloc.split(':')[1])


def split_netloc(netloc):
    return get_hostname(netloc).replace('*', ''), get_port(netloc)


def parse_listen(value):
    for netloc in value.split(','):
        yield split_netloc(netloc)


def parse_uri(uri):
    return urllib.parse.urlparse(uri)


class TrytonConfigParser(configparser.ConfigParser):

    def __init__(self):
        super().__init__(interpolation=None)
        self.add_section('web')
        self.set('web', 'listen', 'localhost:8000')
        self.set('web', 'root',
            os.environ.get('TRYTOND_WEB_ROOT',
                os.path.join(os.path.expanduser('~'), 'www')))
        self.set('web', 'num_proxies', '0')
        self.set('web', 'cache_timeout', str(60 * 60 * 12))
        self.add_section('database')
        self.set('database', 'uri',
            os.environ.get('TRYTOND_DATABASE_URI', 'sqlite://'))
        self.set('database', 'path', os.path.join(
                os.path.expanduser('~'), 'db'))
        self.set('database', 'list', 'True')
        self.set('database', 'retry', '5')
        self.set('database', 'language', 'en')
        self.set('database', 'timeout', str(30 * 60))
        self.add_section('request')
        self.set('request', 'max_size', str(2 * 1024 * 1024))
        self.set('request', 'max_size_authenticated',
            str(2 * 1024 * 1024 * 1024))
        self.add_section('cache')
        self.set('cache', 'transaction', '10')
        self.set('cache', 'model', '200')
        self.set('cache', 'record', '2000')
        self.set('cache', 'field', '100')

        # AKE: cache config from env vars
        self.set('cache', 'class', os.environ.get('TRYTOND_CACHE_CLASS', ''))
        self.set('cache', 'uri', os.environ.get('TRYTOND_CACHE_URI', ''))
        self.set('cache', 'coog_cache_size', '1024')

        self.add_section('report')
        self.set('report', 'unoconv_retry', '2')

        self.add_section('queue')
        self.set('queue', 'worker', 'False')
        self.add_section('ssl')
        self.add_section('email')
        self.set('email', 'uri', 'smtp://localhost:25')
        self.add_section('session')
        self.set('session', 'authentications', 'password')
        self.set('session', 'max_age', str(60 * 60 * 24 * 30))
        self.set('session', 'timeout', str(60 * 5))
        self.set('session', 'max_attempt', '5')
        self.set('session', 'max_attempt_ip_network', '300')
        self.set('session', 'ip_network_4', '32')
        self.set('session', 'ip_network_6', '56')
        self.add_section('password')
        self.set('password', 'length', '8')
        self.set('password', 'reset_timeout', str(24 * 60 * 60))
        self.add_section('bus')
        self.set('bus', 'allow_subscribe', 'False')
        self.set('bus', 'long_polling_timeout', str(5 * 60))
        self.set('bus', 'cache_timeout', '5')
        self.set('bus', 'select_timeout', '5')
        self.add_section('html')
        self.update_environ()
        self.update_etc()

    def update_environ(self):
        for key, value in os.environ.items():
            if not key.startswith('TRYTOND_'):
                continue
            try:
                section, option = key[len('TRYTOND_'):].lower().split('__', 1)
            except ValueError:
                continue
            if section.startswith('wsgi_'):
                section = section.replace('wsgi_', 'wsgi ')
            if not self.has_section(section):
                self.add_section(section)
            self.set(section, option, value)

    def update_etc(self, configfile=os.environ.get('TRYTOND_CONFIG')):
        if isinstance(configfile, str):
            configfile = [configfile]
        if not configfile or not [_f for _f in configfile if _f]:
            return []
        configfile = [os.path.expanduser(filename) for filename in configfile]
        read_files = self.read(configfile)
        logger.info('using %s as configuration files', ', '.join(read_files))
        if configfile != read_files:
            logger.error('could not load %s',
                ','.join(set(configfile) - set(read_files)))
        return configfile

    def get(self, section, option, *args, **kwargs):
        default = kwargs.pop('default', None)
        try:
            return configparser.RawConfigParser.get(self, section, option,
                *args, **kwargs)
        except (configparser.NoOptionError, configparser.NoSectionError):
            return default

    def getint(self, section, option, *args, **kwargs):
        default = kwargs.pop('default', None)
        try:
            return configparser.RawConfigParser.getint(self, section, option,
                *args, **kwargs)
        except (configparser.NoOptionError, configparser.NoSectionError,
                TypeError):
            return default

    def getfloat(self, section, option, *args, **kwargs):
        default = kwargs.pop('default', None)
        try:
            return configparser.RawConfigParser.getfloat(self, section, option,
                *args, **kwargs)
        except (configparser.NoOptionError, configparser.NoSectionError,
                TypeError):
            return default

    def getboolean(self, section, option, *args, **kwargs):
        default = kwargs.pop('default', None)
        try:
            return configparser.RawConfigParser.getboolean(
                self, section, option, *args, **kwargs)
        except (configparser.NoOptionError, configparser.NoSectionError,
                AttributeError):
            return default


config = TrytonConfigParser()

if os.path.basename(getattr(main, '__file__', '')) != 'trytond-stat':
    status.start(config.get('database', 'path'))
